### PR TITLE
rego context aware

### PR DESCRIPTION
**Pull request by flavio**
This PR allows Rego policies to be context aware.

No change to the API of the crate is done. No special action must be done by the consumers of this library.

Fixes: https://github.com/kubewarden/policy-evaluator/issues/353

### CODE REVIEW

**Comment by flavio** [comment]
I've rebased against the `main` branch to include the changes introduced by @fabriziosestito

File: src/runtimes/rego/runtime.rs [hunk 1]
```diff
@@ -1,51 +1,56 @@
...
-                }),
-                json!(settings),
-            ),
+                });
+
+                // OPA data seems to be free-form, except for the
+                // Kubernetes context aware data that must be under the
+                // `kubernetes` key
+                // We don't know the data that is provided by the users via
+                // their settings, hence set the context aware data last, to
```
Comment (viccuad) [inline src/runtimes/rego/runtime.rs:39]:
```suggestion
                // their settings, hence set the context aware data, to
```

File: src/runtimes/rego/gatekeeper_inventory.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
+/// This file defins structs that contain the kubernetes context aware data in a format that is
```
Comment (viccuad) [inline src/runtimes/rego/gatekeeper_inventory.rs:1]:
```suggestion
/// This file defines structs that contain the kubernetes context aware data in a format that is
```

File: src/runtimes/rego/gatekeeper_inventory.rs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+///
+/// If you don't care about the process, jump straight to the section about the inventory.
+///
+/// ## The `inventory` object
+///
+/// As documented [here](https://open-policy-agent.github.io/gatekeeper/website/docs/sync/), the Kubernetes details are made available to all the policies via the `data.inventory` object.
+///
+/// Inventory is a JSON dictionary built using the rules mentioned by the doc:
+/// - For cluster-scoped objects: `data.inventory.cluster[<groupVersion>][<kind>][<name>]`
+/// -  For namespace-scoped objects: `data.inventory.namespace[<namespace>][groupVersion][<kind>][<name>]`
```
Comment (viccuad) [inline src/runtimes/rego/gatekeeper_inventory.rs:12]:
```suggestion
/// - For namespace-scoped objects: `data.inventory.namespace[<namespace>][groupVersion][<kind>][<name>]`
```

File: src/runtimes/rego/opa_inventory.rs [hunk 4]
```diff
@@ -0,0 +1,276 @@
+/// This file builds a the context data required by OPA polices.
+/// ## Docs references
+///
+/// We define OPA policies the ones deployed via [kube-mgmt](https://github.com/open-policy-agent/kube-mgmt), which is an alternative to gatekeeper.
```
Comment (viccuad) [inline src/runtimes/rego/opa_inventory.rs:4]:
```suggestion
/// We define OPA policies as the ones deployed via [kube-mgmt](https://github.com/open-policy-agent/kube-mgmt), which is an alternative to gatekeeper.
```

File: src/runtimes/rego/opa_inventory.rs [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+///
+/// kube-mgmt can be configured to expose Kubernetes resources to the policies. This is described in detail [here](https://github.com/open-policy-agent/kube-mgmt#caching).
+///
+/// By default, the Kubernetes information are made available to the policies inside of `data.kubernetes`. The `kubernetes` key can be changed by the user via a configuration flag, but I think we can ignore this detail and conform to the default behavior.
```
Comment (viccuad) [inline src/runtimes/rego/opa_inventory.rs:8]:
```suggestion
/// By default, the Kubernetes information is made available to the policies inside of `data.kubernetes`. The `kubernetes` key can be changed by the user via a configuration flag, but I think we can ignore this detail and conform to the default behavior.
```

File: src/runtimes/rego/opa_inventory.rs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+///
+/// ## Incoming JSON structure
+///
+/// Kubernetes resources are exposed using this format:
+///
+/// * namespaced resources: `kubernetes.<resource>.<namespace>.<name>`
+/// * cluster wide resources: `kubernetes.<resource>.<name>`
+///
+/// It's important to point out that `<resource>` is the Kubernetes name of the resource obtained when doing:
```
Comment (viccuad) [inline src/runtimes/rego/opa_inventory.rs:17]:
```suggestion
/// It's important to point out that `<resource>` is the Kubernetes plural name of the resource obtained when doing:
```

**Review by viccuad — APPROVED** [review]
LGTM! Nice :).

Added some docstrings quibbles, feel free to batch them in a single commit through the UI.

File: src/runtimes/rego/opa_inventory.rs (hunk 5 above)
Comment (fabriziosestito) [inline src/runtimes/rego/opa_inventory.rs:8]: > but I think we can ignore this detail and conform to the default behavior.

we can remove this sentence

File: src/runtimes/rego/runtime.rs [hunk 7]
```diff
@@ -21,17 +18,39 @@ impl<'a> Runtime<'a> {
...
+        ctx_data: &context_aware::KubernetesContext,
     ) -> AdmissionResponse {
         let uid = request.uid();
 
         // OPA and Gatekeeper expect arguments in different ways. Provide the ones that each expect.
         let (document_to_evaluate, data) = match self.0.policy_execution_mode {
-            RegoPolicyExecutionMode::Opa => (
-                json!({
+            RegoPolicyExecutionMode::Opa => {
+                // Policies for OPA expect the whole `AdmissionReview`
```
Comment (fabriziosestito) [inline src/runtimes/rego/runtime.rs:28]: we can remove this comment as it is not relevant anymore.
the OPA policy expect just a "request" key containing the request object.

**Review by fabriziosestito — APPROVED** [review]
LGTM! thanks

File: src/runtimes/rego/opa_inventory.rs (hunk 5 above)
Comment (flavio) [reply src/runtimes/rego/opa_inventory.rs:8]: I rephrased it to:
> By default, the Kubernetes information are made available to the policies inside of `data.kubernetes`.
> The `kubernetes` key can be changed by the user via a configuration flag, but we decided to conform to the default behavior.

File: src/runtimes/rego/runtime.rs (hunk 7 above)
Comment (flavio) [reply src/runtimes/rego/runtime.rs:28]: removed

File: src/runtimes/rego/opa_inventory.rs (hunk 5 above)
Comment (flavio) [reply src/runtimes/rego/opa_inventory.rs]: fixed

**Pull request merged**